src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -750,5 +733,311 @@ private static string ToHtmlCase(string name)
+                {
+                    valueStart = _index;
+                    var valueEndIndex = _requiredAttributes.IndexOfAny(InvalidCSSQuotelessValueCharacters, _index);
+                    valueEnd = valueEndIndex == -1 ? _requiredAttributes.Length : valueEndIndex;
+                    _index = valueEnd;
+                }
+
+                var value = _requiredAttributes.Substring(valueStart, valueEnd - valueStart);
+
+                return value;
+            }
+
+            private TagHelperRequiredAttributeDescriptor ParseCSSSelector(ErrorSink errorSink)
+            {
+                Debug.Assert(At('['));
+
+                // Move past '['.
+                Next();
+                PassOptionalWhitespace();
+
+                var attributeName = ParseCSSAttributeName(errorSink);
+
+                if (attributeName == null ||
+                    !ValidateName(attributeName, targetingAttributes: true, errorSink: errorSink))
+                {
+                    // Couldn't parse a valid attribute name.
+                    return null;
+                }
+
+                PassOptionalWhitespace();
+
+                if (!EnsureNotAtEnd(errorSink))
+                {
+                    return null;
+                }
+
+                var valueOperator = ParseCSSValueOperator(errorSink);
+
+                PassOptionalWhitespace();
+
+                if (!EnsureNotAtEnd(errorSink))
+                {
+                    return null;
+                }
+
+                var value = ParseCSSValue(errorSink);
+
+                if (value == null)
+                {
+                    // Couldn't parse value
+                    return null;
+                }
+
+                PassOptionalWhitespace();
+
+                if (At(']'))
+                {
+                    // Move past the ending bracket.
+                    Next();
+                }
+                else
+                {
+                    errorSink.OnError(
+                        SourceLocation.Zero,
+                        Resources.TagHelperDescriptorFactory_CouldNotFindMatchingEndBrace,
+                        length: 0);
+                    return null;
+                }
+
+                return new TagHelperRequiredAttributeDescriptor
+                {
+                    Name = attributeName,
+                    Value = value,
+                    Operator = valueOperator,
+                    IsCssSelector = true,
+                };
+            }
+
+            private bool EnsureNotAtEnd(ErrorSink errorSink)
+            {
+                if (AtEnd)
+                {
+                    errorSink.OnError(
+                        SourceLocation.Zero,
+                        Resources.TagHelperDescriptorFactory_CouldNotFindMatchingEndBrace,
+                        length: 0);
+
+                    return false;
+                }
+
+                return true;
+            }
+
+            private void Next()
```

Comment:
I _think_ the position is now checked carefully after every call to this method. But it's very scary that it will stick at `_requiredAttributes.Length`. Looks very much like this factoring leads to fragile code and extra checks everywhere it's used.